### PR TITLE
[clang] Introduce `SemaCUDA`

### PR DESCRIPTION
**Pull request by Endilll**
This patch moves CUDA-related `Sema` function into new `SemaCUDA` class, following the recent example of SYCL, OpenACC, and HLSL. This is a part of the effort to split Sema. Additional context can be found in https://github.com/llvm/llvm-project/pull/82217, https://github.com/llvm/llvm-project/pull/84184, https://github.com/llvm/llvm-project/pull/87634.

### CODE REVIEW

File: clang/include/clang/Sema/SemaCUDA.h [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+  }
+
+  static bool isEqual(const FunctionDeclAndLoc &LHS,
+                      const FunctionDeclAndLoc &RHS) {
+    return LHS.FD == RHS.FD && LHS.Loc == RHS.Loc;
+  }
+};
+} // namespace llvm
+
+#endif // LLVM_CLANG_SEMA_SEMACUDA_H
```
Comment (AaronBallman) [inline clang/include/clang/Sema/SemaCUDA.h:304]: Please add a newline to the end of the file.

**Review by AaronBallman — COMMENTED** [review]
Generally LGTM, though I'd like someone more familiar with CUDA to sign off as well.

**Comment by jhuber6** [comment]
I definitely like moving things out of Sema, since it's so big it always brings my LSP to its knees. Have you run the CUDA / HIP tests in the external test suite or anything?

File: clang/include/clang/Sema/SemaCUDA.h (hunk 1 above)
Comment (Endilll) [reply clang/include/clang/Sema/SemaCUDA.h]: I always forget about it. Thank you!

**Comment by Endilll** [comment]
> Have you run the CUDA / HIP tests in the external test suite or anything?

No, just `check-clang` target. You think it's necessary to run something else before merging? If so, you should point me out to what to run, and preferably give some instructions, as CUDA development is completely foreign to me. Another option is to land and watch out for CUDA bots. I saw we have some, but not sure if they are in a good state.

**Comment by jhuber6** [comment]
> > Have you run the CUDA / HIP tests in the external test suite or anything?
> 
> No, just `check-clang` target. You think it's necessary to run something else before merging? If so, you should point me out to what to run, and preferably give some instructions, as CUDA development is completely foreign to me. Another option is to land and watch out for CUDA bots. I saw we have some, but not sure if they are in a good state.

Yeah, might be sufficient. I don't expect anything to break in non-obvious ways from this.

**Comment by jhuber6** [comment]
Went ahead and applied this locally. Runs some trivial CUDA / HIP programs so I'd say it's fine. I'll wait until @Artem-B or @yxsamliu chime in however as they're the main contributors to this part.

**Review by Artem-B — APPROVED** [review]
LGTM. The changes appear to be mechanical in nature, so `check clang` tests should be sufficient to verify we've re-connected things correctly.

**Comment by fhahn** [comment]
Is it possible that this broke this bot failing with the error below?

```
usr/local/clang-17.0.2/bin/clang++ -DGTEST_HAS_RTTI=0 -D_CINDEX_LIB_ -D_DEBUG -D_GLIBCXX_ASSERTIONS -D_LARGE_FILE_API -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/home/powerllvm/powerllvm_env/aix-ppc64/clang-ppc64-aix/build/tools/clang/tools/libclang -I/home/powerllvm/powerllvm_env/aix-ppc64/clang-ppc64-aix/llvm-project/clang/tools/libclang -I/home/powerllvm/powerllvm_env/aix-ppc64/clang-ppc64-aix/llvm-project/clang/include -I/home/powerllvm/powerllvm_env/aix-ppc64/clang-ppc64-aix/build/tools/clang/include -I/home/powerllvm/powerllvm_env/aix-ppc64/clang-ppc64-aix/build/include -I/home/powerllvm/powerllvm_env/aix-ppc64/clang-ppc64-aix/llvm-project/llvm/include -mcmodel=large -fPIC -Werror -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -pedantic -Wno-long-long -Wc++98-compat-extra-semi -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffunction-sections -fdata-sections -fno-common -Woverloaded-virtual -Wno-nested-anon-types -O3 -DNDEBUG -std=c++17 -fPIC  -fno-exceptions -funwind-tables -fno-rtti -UNDEBUG -MD -MT tools/clang/tools/libclang/CMakeFiles/libclang.dir/CIndexCodeCompletion.cpp.o -MF tools/clang/tools/libclang/CMakeFiles/libclang.dir/CIndexCodeCompletion.cpp.o.d -o tools/clang/tools/libclang/CMakeFiles/libclang.dir/CIndexCodeCompletion.cpp.o -c /home/powerllvm/powerllvm_env/aix-ppc64/clang-ppc64-aix/llvm-project/clang/tools/libclang/CIndexCodeCompletion.cpp
In file included from /home/powerllvm/powerllvm_env/aix-ppc64/clang-ppc64-aix/llvm-project/clang/tools/libclang/CIndexCodeCompletion.cpp:30:
In file included from /home/powerllvm/powerllvm_env/aix-ppc64/clang-ppc64-aix/llvm-project/clang/include/clang/Sema/Sema.h:41:
/home/powerllvm/powerllvm_env/aix-ppc64/clang-ppc64-aix/llvm-project/clang/include/clang/Basic/Cuda.h:59:3: error: expected identifier
   59 |   SM_32,
      |   ^
/usr/include/sys/mac.h:79:15: note: expanded from macro 'SM_32'
   79 | #define SM_32 8         /* number of 32 bit words for markings */
      |               ^
```
 https://lab.llvm.org/buildbot/#/builders/214/builds/11887

**Comment by Endilll** [comment]
@fhahn Yeah, that's theoretically possible. I'll look into it today.

**Comment by Endilll** [comment]
I think I exposed a name conflict with a system header on AIX by including `Cuda.h` in `Sema.h`. Given that this name conflict potentially affects both enum definition and its usage, how should we resolve this?

**Pull request merged**